bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/ChannelPool.java

```diff
@@ -77,19 +85,49 @@ public ChannelPool(List<HeaderInterceptor> headerInterceptors, ChannelFactory fa
    */
   public void ensureChannelCount(int capacity) throws IOException {
     if (totalSize.get() < capacity) {
-      synchronized(this){
+      synchronized (totalSize) {
         if (totalSize.get() < capacity) {
-          List<Channel> newChannelList = new ArrayList<>(channels.get());
-          while(newChannelList.size() < capacity) {
-            newChannelList.add(factory.create());
-          }
-          setChannels(newChannelList);
+          channels.set(
+              ImmutableList.<Channel> builder()
+                  .addAll(channels.get())
+                  .addAll(createNewChannels(capacity - totalSize.get()))
+                  .build()
+          );
           totalSize.set(capacity);
```

Comment:
when is totalSize different from channels.size()?  Should ChannelPool constructor take a capacity parameter? Or at least initialize with ensureChannelCount(1)